pkg/controllers/vdb/sandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,197 @@
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package vdb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/sandboxsc"
+	"github.com/vertica/vertica-kubernetes/pkg/vdbstatus"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SandboxSubclusterReconciler will add subclusters to sandboxes
+type SandboxSubclusterReconciler struct {
+	VRec        *VerticaDBReconciler
+	Log         logr.Logger
+	Vdb         *vapi.VerticaDB // Vdb is the CRD we are acting on
+	PFacts      *PodFacts
+	InitiatorIP string // The IP of the pod that we run vclusterOps from
+	Dispatcher  vadmin.Dispatcher
+	client.Client
+}
+
+// MakeSandboxSubclusterReconciler will build a SandboxSubclusterReconciler object
+func MakeSandboxSubclusterReconciler(vdbrecon *VerticaDBReconciler, log logr.Logger, vdb *vapi.VerticaDB,
+	pfacts *PodFacts, dispatcher vadmin.Dispatcher, cli client.Client) controllers.ReconcileActor {
+	return &SandboxSubclusterReconciler{
+		VRec:       vdbrecon,
+		Log:        log.WithName("SandboxSubclusterReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+		Client:     cli,
+	}
+}
+
+// Reconcile will add subclusters to sandboxes if we found any qualified subclusters
+func (s *SandboxSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op for ScheduleOnly init policy or enterprise db or no sandboxes
+	if s.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly ||
+		!s.Vdb.IsEON() || len(s.Vdb.Spec.Sandboxes) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// We need to collect pod facts for finding qualified subclusters
+	if err := s.PFacts.Collect(ctx, s.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return s.sandboxSubclusters(ctx)
+}
+
+// sandboxSubclusters will add subclusters to their sandboxes defined in the vdb
+func (s *SandboxSubclusterReconciler) sandboxSubclusters(ctx context.Context) (ctrl.Result, error) {
+	// find qualified subclusters with their sandboxes
+	scSbMap, err := s.fetchSubclustersWithSandboxes()
+	// the error indicates we have down nodes in a subcluster that will be sandboxed.
+	// we simply requeue the iteration for waiting the nodes to be up.
+	if err != nil {
+		s.Log.Info("Requeue because a pod is not ready", "details", err.Error())
+		return ctrl.Result{}, nil
+	}
```

Comment:
Alex told me we couldn't sandbox a subcluster if the nodes were down because down nodes cannot create a check point. This is a server limitation. We might need to review the sandbox design.